File: CurrencyConverter.js
const readline = require('readline').createInterface({
    input: process.stdin,
    output: process.stdout
});

const fs = require('fs');

const currencies = ['USD', 'BAM', 'EUR', 'CHF', 'AUD'];
const historyFile = 'conversion_history.json';
const presetsFile = 'preset_values.json';
const helpFile = 'help_text.json';

const question = (query) => new Promise((resolve) => readline.question(query, resolve));


const sleep = (ms) => new Promise(resolve => setTimeout(resolve, ms));
const clearScreen = () => console.clear();

async function waitForUser() {
    await question("\nPress Enter to continue...");
    clearScreen();
}

async function loadingAnimation() {
    const spinnerChars = ['⠋', '⠙', '⠹', '⠸', '⠼', '⠴', '⠦', '⠧', '⠇', '⠏'];
    let spinnerIndex = 0;

    console.log("\n=== Initializing Currency Converter ===\n");

    for (const currency of currencies) {
        process.stdout.write(`Loading ${currency} `);
        for (let i = 0; i < 20; i++) {
            process.stdout.write(`\r${currency}: [${spinnerChars[spinnerIndex]}] Loading... `);
            spinnerIndex = (spinnerIndex + 1) % spinnerChars.length;
            await sleep(100);
        }
        process.stdout.write(`\r${currency}: [✓] Loaded successfully!   \n`);
    }

    console.log("\nSystem initialization complete!");
    console.log("Starting currency converter...\n");
    await sleep(1000);
    clearScreen();
}

function createHelpFile() {
    if (!fs.existsSync(helpFile)) {
        const helpContent = {
            "conversion": {
                "title": "How Currency Conversion Works",
                "steps": [
                    "1. Select the currency you want to convert FROM (e.g., USD, EUR)",
                    "2. Select the currency you want to convert TO",
                    "3. Enter the amount you want to convert",
                    "4. The system will use current exchange rates to calculate the result"
                ],
                "notes": [
                    "• All conversions are saved in your history",
                    "• Exchange rates are updated regularly",
                    "• You can view your conversion history anytime"
                ],
                "example": "Example: Converting 100 USD to EUR\n- Enter 'USD' as source currency\n- Enter 'EUR' as target currency\n- Enter '100' as amount\n- System will show result (e.g., 92.00 EUR)"
            },
            "predictions": {
                "title": "Understanding Predictions",
                "daily": [
                    "Daily predictions show possible rate changes over 24 hours",
                    "Predictions use current rates with market volatility factors",
                    "All major currency pairs are included in daily predictions"
                ],
                "hourly": [
                    "Hourly predictions show detailed short-term trends",
                    "Choose between 1, 3, or 6-hour prediction windows",
                    "Includes trend indicators (↑/↓) and percentage changes"
                ]
            }
        };
        fs.writeFileSync(helpFile, JSON.stringify(helpContent, null, 2));
    }
}

function loadPresets() {
    try {
        return JSON.parse(fs.readFileSync(presetsFile, 'utf8'));
    } catch (error) {
        const presetRates = {
            'USD': {'BAM': 1.85, 'EUR': 0.92, 'CHF': 0.89, 'AUD': 1.53},
            'BAM': {'USD': 0.54, 'EUR': 0.51, 'CHF': 0.48, 'AUD': 0.83},
            'EUR': {'USD': 1.09, 'BAM': 1.96, 'CHF': 0.97, 'AUD': 1.67},
            'CHF': {'USD': 1.12, 'BAM': 2.08, 'EUR': 1.03, 'AUD': 1.72},
            'AUD': {'USD': 0.65, 'BAM': 1.21, 'EUR': 0.60, 'CHF': 0.58}
        };
        fs.writeFileSync(presetsFile, JSON.stringify(presetRates, null, 2));
        return presetRates;
    }
}

function showHelp() {
    try {
        const helpContent = JSON.parse(fs.readFileSync(helpFile, 'utf8'));
        
        console.log("\n=== Currency Converter Help ===");
        
        console.log(`\n${helpContent.conversion.title}`);
        console.log("\nSteps:");
        helpContent.conversion.steps.forEach(step => console.log(step));
        
        console.log("\nImportant Notes:");
        helpContent.conversion.notes.forEach(note => console.log(note));
        
        console.log("\n" + helpContent.conversion.example);
        
        console.log(`\n${helpContent.predictions.title}`);
        console.log("\nDaily Predictions:");
        helpContent.predictions.daily.forEach(info => console.log(`• ${info}`));
        
        console.log("\nHourly Predictions:");
        helpContent.predictions.hourly.forEach(info => console.log(`• ${info}`));
    } catch (error) {
        console.log("\nHelp information not available.");
    }
}

function saveConversion(fromCurr, toCurr, amount, result) {
    let history = [];
    try {
        history = JSON.parse(fs.readFileSync(historyFile, 'utf8'));
    } catch (error) {
    }

    history.push({
        date: new Date().toISOString().replace('T', ' ').slice(0, 19),
        from: fromCurr,
        to: toCurr,
        amount: amount,
        result: result
    });

    fs.writeFileSync(historyFile, JSON.stringify(history, null, 2));
}

function showPopularConversions(rates) {
    const commonPairs = [
        ['USD', 'BAM'],
        ['EUR', 'BAM'],
        ['USD', 'EUR'],
        ['CHF', 'EUR'],
        ['AUD', 'USD']
    ];
    
    const totalConversions = Math.floor(Math.random() * (1000 - 500) + 500);
    console.log("\n=== Most Used Conversions Today ===");
    console.log(`Total conversions today: ${totalConversions}\n`);
    console.log("Currency Pair     Current Rate    Times Used");
    console.log("-".repeat(45));
    
    commonPairs.forEach(([fromCurr, toCurr]) => {
        const conversions = Math.floor(totalConversions * (Math.random() * 0.3 + 0.05));
        const rate = rates[fromCurr][toCurr];
        console.log(`${fromCurr}/${toCurr.padEnd(8)}     ${rate.toFixed(4)}          ${conversions}`);
    });
}

function generatePredictions(rates) {
    const predictions = {};
    currencies.forEach(base => {
        predictions[base] = {};
        currencies.forEach(target => {
            if (base !== target) {
                const currentRate = rates[base][target];
                const variation = (Math.random() * 0.1) - 0.05;
                predictions[base][target] = Number((currentRate * (1 + variation)).toFixed(4));
            }
        });
    });
    return predictions;
}

function generateHourlyPredictions(rates, fromCurr, toCurr, hours) {
    const baseRate = rates[fromCurr][toCurr];
    const predictions = [];
    const currentHour = new Date();
    
    const volatility = 0.02;
    let prevVariation = 0;
    
    for (let i = 0; i < hours; i++) {
        const nextHour = new Date(currentHour);
        nextHour.setHours(currentHour.getHours() + i);
        
        const variation = (Math.random() * volatility * 2 - volatility) + (prevVariation * 0.3);
        const predictedRate = baseRate * (1 + variation);
        
        predictions.push({
            hour: nextHour.getHours().toString().padStart(2, '0') + ':00',
            rate: Number(predictedRate.toFixed(4)),
            change: Number((variation * 100).toFixed(2)),
            trend: variation > 0 ? '↑' : '↓'
        });
        
        prevVariation = variation;
    }
    
    return predictions;
}

async function handleHourlyPredictions(rates) {
    console.log("\nAvailable currencies:", currencies.join(', '));
    
    const fromCurr = (await question("Convert from (currency code): ")).toUpperCase();
    const toCurr = (await question("Convert to (currency code): ")).toUpperCase();
    
    if (!currencies.includes(fromCurr) || !currencies.includes(toCurr)) {
        console.log("Invalid currency code!");
        return;
    }
    
    let hours;
    while (true) {
        hours = parseInt(await question("\nEnter number of hours to predict (1, 3, or 6): "));
        if ([1, 3, 6].includes(hours)) break;
        console.log("Please enter either 1, 3, or 6 hours.");
    }
    
    const predictions = generateHourlyPredictions(rates, fromCurr, toCurr, hours);
    const currentRate = rates[fromCurr][toCurr];
    
    console.log(`\n=== ${fromCurr}/${toCurr} Predictions ===`);
    console.log(`Current rate: ${currentRate.toFixed(4)}`);
    console.log(`\nPredicted rates for next ${hours} hour(s):`);
    console.log("\nTime    Rate      Change   Trend");
    console.log("-".repeat(35));
    predictions.forEach(pred => {
        console.log(`${pred.hour}   ${pred.rate.toFixed(4)}   ${pred.change.toFixed(2).padStart(6)}%   ${pred.trend}`);
    });
}

async function handleConversion(rates) {
    console.log("\nAvailable currencies:", currencies.join(', '));
    
    const fromCurr = (await question("Convert from (currency code): ")).toUpperCase();
    const toCurr = (await question("Convert to (currency code): ")).toUpperCase();
    
    if (!currencies.includes(fromCurr) || !currencies.includes(toCurr)) {
        console.log("Invalid currency code!");
        return;
    }
    
    const amount = parseFloat(await question("Enter amount: "));
    if (isNaN(amount)) {
        console.log("Invalid amount!");
        return;
    }
    
    const result = amount * rates[fromCurr][toCurr];
    console.log(`\nResult: ${amount} ${fromCurr} = ${result.toFixed(2)} ${toCurr}`);
    saveConversion(fromCurr, toCurr, amount, result);
}

function handlePredictions(rates) {
    const predictions = generatePredictions(rates);
    console.log("\n=== Currency Predictions ===");
    Object.entries(predictions).forEach(([base, targets]) => {
        console.log(`\n${base} predictions:`);
        Object.entries(targets).forEach(([target, rate]) => {
            console.log(`${base} to ${target}: ${rate.toFixed(4)}`);
        });
    });
}

function handleHistory() {
    try {
        const history = JSON.parse(fs.readFileSync(historyFile, 'utf8'));
        if (history.length === 0) {
            console.log("\nNo conversion history found.");
        } else {
            console.log("\n=== Conversion History ===");
            history.forEach(entry => {
                console.log(`${entry.date}: ${entry.amount} ${entry.from} = ${entry.result.toFixed(2)} ${entry.to}`);
            });
        }
    } catch (error) {
        console.log("\nNo conversion history found.");
    }
}

function clearHistory() {
    fs.writeFileSync(historyFile, JSON.stringify([]));
}

async function showMenu() {
    clearScreen();
    console.log("=== Currency Converter Menu ===");
    console.log("1. Convert Currency");
    console.log("2. Hourly Predictions");
    console.log("3. Daily Predictions");
    console.log("4. Popular Conversions");
    console.log("5. View Conversion History");
    console.log("6. Clear History");
    console.log("7. Help");
    console.log("8. Exit");
    
    return await question("\nEnter your choice (1-8): ");
}

async function main() {
    try {
        await loadingAnimation();
        createHelpFile();
        const rates = loadPresets();
        
        while (true) {
            const choice = await showMenu();
            
            switch (choice) {
                case "1":
                    await handleConversion(rates);
                    break;
                case "2":
                    await handleHourlyPredictions(rates);
                    break;
                case "3":
                    handlePredictions(rates);
                    break;
                case "4":
                    showPopularConversions(rates);
                    break;
                case "5":
                    handleHistory();
                    break;
                case "6":
                    clearHistory();
                    console.log("\nHistory cleared successfully!");
                    break;
                case "7":
                    showHelp();
                    break;
                case "8":
                    console.log("\nThank you for using Currency Converter!");
                    readline.close();
                    process.exit(0);
                default:
                    console.log("\nInvalid choice! Please try again.");
            }
            await waitForUser();
        }
    } catch (error) {
        console.error("An error occurred:", error);
        readline.close();
    }
}

main().catch(console.error);

File: CurrencyConverter.cs
using System;
using System.Collections.Generic;
using System.Text.Json;
using System.IO;
using System.Linq;
using System.Threading.Tasks;
using System.Diagnostics;
class CurrencyConverter
{
    private static readonly string[] currencies = { "USD", "BAM", "EUR", "CHF", "AUD" };
    private const string HistoryFile = "conversion_history.json";
    private const string PresetsFile = "preset_values.json";
    private const string HelpFile = "help_text.json";
    private static readonly Random random = new Random();
    class ConversionHistory
    {
        public string Date { get; set; }
        public string From { get; set; }
        public string To { get; set; }
        public double Amount { get; set; }
        public double Result { get; set; }
    }
    static async Task LoadingAnimation()
    {
        char[] spinner = { '⠋', '⠙', '⠹', '⠸', '⠼', '⠴', '⠦', '⠧', '⠇', '⠏' };
        int spinnerIndex = 0;
        Console.WriteLine("\n=== Initializing Currency Converter ===\n");   
        foreach (string currency in currencies)
        {
            Console.Write($"Loading {currency} ");
            for (int i = 0; i < 20; i++)
            {
                Console.Write($"\r{currency}: [{spinner[spinnerIndex]}] Loading... ");
                spinnerIndex = (spinnerIndex + 1) % spinner.Length;
                await Task.Delay(100);
            }
            Console.WriteLine($"\r{currency}: [✓] Loaded successfully!   ");
        }
        Console.WriteLine("\nSystem initialization complete!");
        Console.WriteLine("Starting currency converter...\n");
        await Task.Delay(1000);
        ClearScreen();
    }
    static void ClearScreen()
    {
        Console.Clear();
    }
    static void WaitForUser()
    {
        Console.WriteLine("\nPress Enter to continue...");
        Console.ReadLine();
        ClearScreen();
    }
    static Dictionary<string, Dictionary<string, double>> LoadPresets()
    {
        try
        {
            string json = File.ReadAllText(PresetsFile);
            return JsonSerializer.Deserialize<Dictionary<string, Dictionary<string, double>>>(json);
        }
        catch (FileNotFoundException)
        {
            var presetRates = new Dictionary<string, Dictionary<string, double>>
            {
                ["USD"] = new Dictionary<string, double> { ["BAM"] = 1.85, ["EUR"] = 0.92, ["CHF"] = 0.89, ["AUD"] = 1.53 },
                ["BAM"] = new Dictionary<string, double> { ["USD"] = 0.54, ["EUR"] = 0.51, ["CHF"] = 0.48, ["AUD"] = 0.83 },
                ["EUR"] = new Dictionary<string, double> { ["USD"] = 1.09, ["BAM"] = 1.96, ["CHF"] = 0.97, ["AUD"] = 1.67 },
                ["CHF"] = new Dictionary<string, double> { ["USD"] = 1.12, ["BAM"] = 2.08, ["EUR"] = 1.03, ["AUD"] = 1.72 },
                ["AUD"] = new Dictionary<string, double> { ["USD"] = 0.65, ["BAM"] = 1.21, ["EUR"] = 0.60, ["CHF"] = 0.58 }
            };
            string json = JsonSerializer.Serialize(presetRates, new JsonSerializerOptions { WriteIndented = true });
            File.WriteAllText(PresetsFile, json);
            return presetRates;
        }
    }
    static void CreateHelpFile()
    {
        if (!File.Exists(HelpFile))
        {
            var helpContent = new
            {
                conversion = new
                {
                    title = "How Currency Conversion Works",
                    steps = new[]
                    {
                        "1. Select the currency you want to convert FROM (e.g., USD, EUR)",
                        "2. Select the currency you want to convert TO",
                        "3. Enter the amount you want to convert",
                        "4. The system will use current exchange rates to calculate the result"
                    },
                    notes = new[]
                    {
                        "• All conversions are saved in your history",
                        "• Exchange rates are updated regularly",
                        "• You can view your conversion history anytime"
                    },
                    example = "Example: Converting 100 USD to EUR\n- Enter 'USD' as source currency\n- Enter 'EUR' as target currency\n- Enter '100' as amount\n- System will show result (e.g., 92.00 EUR)"
                },
                predictions = new
                {
                    title = "Understanding Predictions",
                    daily = new[]
                    {
                        "Daily predictions show possible rate changes over 24 hours",
                        "Predictions use current rates with market volatility factors",
                        "All major currency pairs are included in daily predictions"
                    },
                    hourly = new[]
                    {
                        "Hourly predictions show detailed short-term trends",
                        "Choose between 1, 3, or 6-hour prediction windows",
                        "Includes trend indicators (↑/↓) and percentage changes"
                    }
                }
            };

            string json = JsonSerializer.Serialize(helpContent, new JsonSerializerOptions { WriteIndented = true });
            File.WriteAllText(HelpFile, json);
        }
    }

    static void ShowHelp()
    {
        try
        {
            string json = File.ReadAllText(HelpFile);
            using JsonDocument doc = JsonDocument.Parse(json);
            var root = doc.RootElement;

            Console.WriteLine("\n=== Currency Converter Help ===");

            // Show conversion help
            Console.WriteLine($"\n{root.GetProperty("conversion").GetProperty("title").GetString()}");
            Console.WriteLine("\nSteps:");
            foreach (var step in root.GetProperty("conversion").GetProperty("steps").EnumerateArray())
            {
                Console.WriteLine(step.GetString());
            }

            Console.WriteLine("\nImportant Notes:");
            foreach (var note in root.GetProperty("conversion").GetProperty("notes").EnumerateArray())
            {
                Console.WriteLine(note.GetString());
            }

            Console.WriteLine("\n" + root.GetProperty("conversion").GetProperty("example").GetString());

            // Show predictions help
            Console.WriteLine($"\n{root.GetProperty("predictions").GetProperty("title").GetString()}");
            Console.WriteLine("\nDaily Predictions:");
            foreach (var info in root.GetProperty("predictions").GetProperty("daily").EnumerateArray())
            {
                Console.WriteLine($"• {info.GetString()}");
            }

            Console.WriteLine("\nHourly Predictions:");
            foreach (var info in root.GetProperty("predictions").GetProperty("hourly").EnumerateArray())
            {
                Console.WriteLine($"• {info.GetString()}");
            }
        }
        catch (FileNotFoundException)
        {
            Console.WriteLine("\nHelp information not available.");
        }
    }

    static void SaveConversion(string fromCurr, string toCurr, double amount, double result)
    {
        List<ConversionHistory> history;
        try
        {
            string json = File.ReadAllText(HistoryFile);
            history = JsonSerializer.Deserialize<List<ConversionHistory>>(json);
        }
        catch (FileNotFoundException)
        {
            history = new List<ConversionHistory>();
        }

        history.Add(new ConversionHistory
        {
            Date = DateTime.Now.ToString("yyyy-MM-dd HH:mm:ss"),
            From = fromCurr,
            To = toCurr,
            Amount = amount,
            Result = result
        });

        string newJson = JsonSerializer.Serialize(history, new JsonSerializerOptions { WriteIndented = true });
        File.WriteAllText(HistoryFile, newJson);
    }

    static void ShowPopularConversions(Dictionary<string, Dictionary<string, double>> rates)
    {
        var commonPairs = new[]
        {
            ("USD", "BAM"),
            ("EUR", "BAM"),
            ("USD", "EUR"),
            ("CHF", "EUR"),
            ("AUD", "USD")
        };

        int totalConversions = random.Next(500, 1001);
        Console.WriteLine("\n=== Most Used Conversions Today ===");
        Console.WriteLine($"Total conversions today: {totalConversions}\n");
        Console.WriteLine("Currency Pair     Current Rate    Times Used");
        Console.WriteLine("-".PadRight(45, '-'));

        foreach (var (from, to) in commonPairs)
        {
            int conversions = (int)(totalConversions * random.NextDouble() * 0.3 + 0.05);
            double rate = rates[from][to];
            Console.WriteLine($"{from}/{to,-8}     {rate:F4}          {conversions}");
        }
    }

    static Dictionary<string, Dictionary<string, double>> GeneratePredictions(Dictionary<string, Dictionary<string, double>> rates)
    {
        var predictions = new Dictionary<string, Dictionary<string, double>>();
        foreach (var baseCurrency in currencies)
        {
            predictions[baseCurrency] = new Dictionary<string, double>();
            foreach (var targetCurrency in currencies)
            {
                if (baseCurrency != targetCurrency)
                {
                    double currentRate = rates[baseCurrency][targetCurrency];
                    double variation = (random.NextDouble() * 0.1) - 0.05;
                    predictions[baseCurrency][targetCurrency] = Math.Round(currentRate * (1 + variation), 4);
                }
            }
        }
        return predictions;
    }

    static List<ConversionHistory> ViewHistory()
    {
        try
        {
            string json = File.ReadAllText(HistoryFile);
            return JsonSerializer.Deserialize<List<ConversionHistory>>(json);
        }
        catch (FileNotFoundException)
        {
            return new List<ConversionHistory>();
        }
    }

    static void ClearHistory()
    {
        File.WriteAllText(HistoryFile, JsonSerializer.Serialize(new List<ConversionHistory>()));
    }

    static List<Dictionary<string, object>> GenerateHourlyPredictions(
        Dictionary<string, Dictionary<string, double>> rates,
        string fromCurr,
        string toCurr,
        int hours)
    {
        double baseRate = rates[fromCurr][toCurr];
        var predictions = new List<Dictionary<string, object>>();
        DateTime currentHour = DateTime.Now;
        
        const double volatility = 0.02;
        double prevVariation = 0;

        for (int i = 0; i < hours; i++)
        {
            DateTime nextHour = currentHour.AddHours(i);
            double variation = (random.NextDouble() * 2 - 1) * volatility + (prevVariation * 0.3);
            double predictedRate = baseRate * (1 + variation);

            predictions.Add(new Dictionary<string, object>
            {
                ["hour"] = nextHour.ToString("HH:00"),
                ["rate"] = Math.Round(predictedRate, 4),
                ["change"] = Math.Round(variation * 100, 2),
                ["trend"] = variation > 0 ? "↑" : "↓"
            });

            prevVariation = variation;
        }

        return predictions;
    }

    static void HandleHourlyPredictions(Dictionary<string, Dictionary<string, double>> rates)
    {
        Console.WriteLine("\nAvailable currencies: " + string.Join(", ", currencies));
        Console.Write("Convert from (currency code): ");
        string fromCurr = Console.ReadLine().ToUpper();
        Console.Write("Convert to (currency code): ");
        string toCurr = Console.ReadLine().ToUpper();

        if (!currencies.Contains(fromCurr) || !currencies.Contains(toCurr))
        {
            Console.WriteLine("Invalid currency code!");
            return;
        }

        int hours;
        while (true)
        {
            Console.Write("\nEnter number of hours to predict (1, 3, or 6): ");
            if (int.TryParse(Console.ReadLine(), out hours) && new[] { 1, 3, 6 }.Contains(hours))
                break;
            Console.WriteLine("Please enter either 1, 3, or 6 hours.");
        }

        var predictions = GenerateHourlyPredictions(rates, fromCurr, toCurr, hours);
        double currentRate = rates[fromCurr][toCurr];

        Console.WriteLine($"\n=== {fromCurr}/{toCurr} Predictions ===");
        Console.WriteLine($"Current rate: {currentRate:F4}");
        Console.WriteLine($"\nPredicted rates for next {hours} hour(s):");
        Console.WriteLine("\nTime    Rate      Change   Trend");
        Console.WriteLine("-".PadRight(35, '-'));

        foreach (var pred in predictions)
        {
            Console.WriteLine($"{pred["hour"]}   {pred["rate"]:F4}   {pred["change"],6:F2}%   {pred["trend"]}");
        }
    }

    static void HandleConversion(Dictionary<string, Dictionary<string, double>> rates)
    {
        Console.WriteLine("\nAvailable currencies: " + string.Join(", ", currencies));
        Console.Write("Convert from (currency code): ");
        string fromCurr = Console.ReadLine().ToUpper();
        Console.Write("Convert to (currency code): ");
        string toCurr = Console.ReadLine().ToUpper();

        if (!currencies.Contains(fromCurr) || !currencies.Contains(toCurr))
        {
            Console.WriteLine("Invalid currency code!");
            return;
        }

        Console.Write("Enter amount: ");
        if (double.TryParse(Console.ReadLine(), out double amount))
        {
            double result = amount * rates[fromCurr][toCurr];
            Console.WriteLine($"\nResult: {amount} {fromCurr} = {result:F2} {toCurr}");
            SaveConversion(fromCurr, toCurr, amount, result);
        }
        else
        {
            Console.WriteLine("Invalid amount!");
        }
    }

    static void HandlePredictions(Dictionary<string, Dictionary<string, double>> rates)
    {
        var predictions = GeneratePredictions(rates);
        Console.WriteLine("\n=== Currency Predictions ===");
        foreach (var baseCurrency in predictions)
        {
            Console.WriteLine($"\n{baseCurrency.Key} predictions:");
            foreach (var (target, rate) in baseCurrency.Value)
            {
                Console.WriteLine($"{baseCurrency.Key} to {target}: {rate:F4}");
            }
        }
    }

    static void HandleHistory()
    {
        var history = ViewHistory();
        
        if (!history.Any())
        {
            Console.WriteLine("\nNo conversion history found.");
        }
        else
        {
            Console.WriteLine("\n=== Conversion History ===");
            foreach (var entry in history)
            {
                Console.WriteLine($"{entry.Date}: {entry.Amount} {entry.From} = {entry.Result:F2} {entry.To}");
            }
        }
    }
    static async Task Main()
    {
        await LoadingAnimation();
        CreateHelpFile();
        var rates = LoadPresets();

        while (true)
        {
            ClearScreen();
            Console.WriteLine("=== Currency Converter Menu ===");
            Console.WriteLine("1. Convert Currency");
            Console.WriteLine("2. Hourly Predictions");
            Console.WriteLine("3. Daily Predictions");
            Console.WriteLine("4. Popular Conversions");
            Console.WriteLine("5. View Conversion History");
            Console.WriteLine("6. Clear History");
            Console.WriteLine("7. Help");
            Console.WriteLine("8. Exit");

            Console.Write("\nEnter your choice (1-8): ");
            string choice = Console.ReadLine();

            switch (choice)
            {
                case "1":
                    HandleConversion(rates);
                    break;
                case "2":
                    HandleHourlyPredictions(rates);
                    break;
                case "3":
                    HandlePredictions(rates);
                    break;
                case "4":
                    ShowPopularConversions(rates);
                    break;
                case "5":
                    HandleHistory();
                    break;
                case "6":
                    ClearHistory();
                    Console.WriteLine("\nHistory cleared successfully!");
                    break;
                case "7":
                    ShowHelp();
                    break;
                case "8":
                    Console.WriteLine("\nThank you for using Currency Converter!");
                    return;
                default:
                    Console.WriteLine("\nInvalid choice! Please try again.");
                    break;
            }
            WaitForUser();
        }
    }
}